### PR TITLE
Add implicit any errors for destructuring computed names which aren't late bound and have no corresponding index

### LINKED ISSUE #23334: Missing implicit any error in destructuring with computed property
**TypeScript Version:**  2.9.0-dev.20180411

**Search Terms:** 

**Code**

```ts
// --noImplicitAny
let named = 'foo';
let {[named]: prop} = {prop: 'foo'};
```

**Expected behavior:**
`Element implicitly has an 'any' type because type '{ prop: string; }' has no index signature.`

**Actual behavior:**
`prop` is `any`, but there is no implicit any error

**Playground Link:** https://agentcooper.github.io/typescript-play/#code/DYUwLgBAdghgtiAJhAvBA5AMwPbfQbgChRIBvAbVgUQF0AuCABwCdtGBfVCUlthrXOnb4gA

**Related Issues:**
https://github.com/Microsoft/TypeScript/pull/23297#issuecomment-380004267

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #23334

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -4173,7 +4173,23 @@ namespace ts {
                     const isLate = isLateBindableName(name);
                     const isWellKnown = isComputedPropertyName(name) && isWellKnownSymbolSyntactically(name.expression);
                     if (!isLate && !isWellKnown && isComputedNonLiteralName(name)) {
-                        return anyType;
+                        const exprType = checkExpression((name as ComputedPropertyName).expression);
+                        let indexerType: Type;
+                        if (isTypeAssignableToKind(exprType, TypeFlags.NumberLike)) {
+                            indexerType = getIndexTypeOfType(parentType, IndexKind.Number);
+                        }
+                        else if (isTypeAssignableToKind(exprType, TypeFlags.StringLike)) {
```
Comment (mhegazy) [inline src/compiler/checker.ts]: string is valid for number indexing as well.. so symbols aside..
`indexrType = (isTypeAssignableToKind(exprType, TypeFlags.NumberLike) && getIndexTypeOfType(parentType, IndexKind.Number) || getIndexTypeOfType(parentType, IndexKind.String)`

File: tests/cases/compiler/lateBoundDestructuringImplicitAnyError.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// @noImplicitAny: true
+let named = "foo";
+let {[named]: prop} = {prop: "foo"};
```
Comment (mhegazy) [inline tests/cases/compiler/lateBoundDestructuringImplicitAnyError.ts]:
```ts
const named = Symbol("");
let { [named]: prop } = strIndexed;
void prop;``
```

today that is `string` ... we probably wanna keep it this way until we have a symbol indexer.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -4173,7 +4173,23 @@ namespace ts {
                     const isLate = isLateBindableName(name);
                     const isWellKnown = isComputedPropertyName(name) && isWellKnownSymbolSyntactically(name.expression);
                     if (!isLate && !isWellKnown && isComputedNonLiteralName(name)) {
-                        return anyType;
+                        const exprType = checkExpression((name as ComputedPropertyName).expression);
+                        let indexerType: Type;
+                        if (isTypeAssignableToKind(exprType, TypeFlags.NumberLike)) {
+                            indexerType = getIndexTypeOfType(parentType, IndexKind.Number) || getIndexTypeOfType(parentType, IndexKind.String);
+                        }
+                        else if (isTypeAssignableToKind(exprType, TypeFlags.StringLike)) {
```
Comment (mhegazy) [inline src/compiler/checker.ts]:
```ts
const named = Symbol("");
let { [named]: prop } = strIndexed;
void prop;``
```

today that is `string` ... we probably wanna keep it this way until we have a symbol indexer.

**Comment by mhegazy** [comment]
We need to update it to handle symbols like we do in regular indexing.

**Comment by weswigham** [comment]
@mhegazy Does this look like the correct errors to issue for symbol destructuring (and should they be noImplicitAny)?

**Pull request merged**